### DISCUSSION TITLE
How to insert raw XML into document?

### DISCUSSION THREAD

**Discussion by chriscrittenden** [dolanmiu/docx#1786] [Q&A]
In Word, you can insert a field (see image). I'm trying to replicate this with docx. This is the XML that results from doing this in word:

<w:r>
    <w:rPr>
        <w:noProof />
    </w:rPr>
    <w:instrText xml:space="preserve"> INCLUDEPICTURE  \d "[URL TO IMAGE]" \* MERGEFORMATINET </w:instrText>
</w:r>

Is there a way to insert this raw XML into my document?

<img width="439" alt="Screen Shot 2022-11-19 at 9 10 55 AM" src="https://user-images.githubusercontent.com/47259885/202860477-e7609158-5824-4d2b-8dbb-f71cbb869557.png">

**Comment by dolanmiu** [comment]
Not possible at the moment, but sounds powerful

AFAIK, they need to be part of fields? is that true?

Related docs:
http://officeopenxml.com/WPfieldInstructions.php
http://officeopenxml.com/WPgeneralFieldSwitches.php

**Reply by chriscrittenden** [reply]
Thanks for the quick reply!

Yes, I believe they need to be a part of fields. Is there a way to insert Fields (described in the first link you sent)?

**Reply by dolanmiu** [reply]
Yup there is, check this out:

https://github.com/dolanmiu/docx/blob/master/demo/66-fields.ts

**Reply by chriscrittenden** [reply]
I'm so close! Thanks for your help.

When I open the file, I have to press f9 for it to fetch the external website. [This SO question](https://stackoverflow.com/questions/25695480/insert-field-codes-into-document-using-docx4java) for the exact same problem with Docx4J (a java package) suggests that you just have to add w:pict (or a w:drawing) after the field separator. Any idea how I can accomplish that here?

**Reply by chriscrittenden** [reply]
[Here's a codepen](https://codepen.io/ccritten/pen/wvXpGLK?editors=1010) showing the problem. When you open the generated document you have to select all and then press f9 to fetch the image.

**Reply by dolanmiu** [reply]
This feature is not implemented yet:

It seems like there is a lot of work to do on it, 

basically this needs to be implemented fully:

http://www.datypic.com/sc/ooxml/s-vml-main.xsd.html

The extract from stack overflow you sent:

```xml
<w:pict>
  <v:shapetype coordsize="21600,21600" filled="f" id="_x0000_t75" o:preferrelative="t" o:spt="75.0" path="m@4@5l@4@11@9@11@9@5xe" stroked="f">
    <v:stroke joinstyle="miter"/>
    <v:formulas>
      <v:f eqn="if lineDrawn pixelLineWidth 0"/>
      <v:f eqn="sum @0 1 0"/>
      <v:f eqn="sum 0 0 @1"/>
      <v:f eqn="prod @2 1 2"/>
      <v:f eqn="prod @3 21600 pixelWidth"/>
      <v:f eqn="prod @3 21600 pixelHeight"/>
      <v:f eqn="sum @0 0 1"/>
      <v:f eqn="prod @6 1 2"/>
      <v:f eqn="prod @7 21600 pixelWidth"/>
      <v:f eqn="sum @8 21600 0"/>
      <v:f eqn="prod @7 21600 pixelHeight"/>
      <v:f eqn="sum @10 21600 0"/>
    </v:formulas>
    <v:path gradientshapeok="t" o:connecttype="rect" o:extrusionok="f"/>
    <o:lock aspectratio="t" v:ext="edit"/>
  </v:shapetype>
  <v:shape id="_x0000_i1025" style="width:100pt;height:100pt" type="#_x0000_t75">
    <v:imagedata r:href="rId6" r:id="rId5"/>
  </v:shape>
</w:pict>
```
